File: examples/java/src/main/java/com/tokenizers/Tokenizer.java
package com.tokenizers;

/**
 * Java wrapper for the tokenizers-sys C bindings
 */
public class Tokenizer implements AutoCloseable {
    
    static {
        // Load the native library
        System.loadLibrary("tokenizers_jni");
    }

    private long handle;

    private Tokenizer(long handle) {
        this.handle = handle;
    }

    /**
     * Create a tokenizer from a pretrained model
     * @param name The model name (e.g., "bert-base-cased", "gpt2")
     * @return A new Tokenizer instance
     */
    public static Tokenizer fromPretrained(String name) {
        long handle = nativeFromPretrained(name, null, null);
        if (handle == 0) {
            throw new RuntimeException("Failed to load tokenizer: " + name);
        }
        return new Tokenizer(handle);
    }

    /**
     * Create a tokenizer from a pretrained model with parameters
     * @param name The model name
     * @param revision The model revision (can be null)
     * @param token The auth token (can be null)
     * @return A new Tokenizer instance
     */
    public static Tokenizer fromPretrained(String name, String revision, String token) {
        long handle = nativeFromPretrained(name, revision, token);
        if (handle == 0) {
            throw new RuntimeException("Failed to load tokenizer: " + name);
        }
        return new Tokenizer(handle);
    }

    /**
     * Create a tokenizer from a file
     * @param path Path to the tokenizer file
     * @return A new Tokenizer instance
     */
    public static Tokenizer fromFile(String path) {
        long handle = nativeFromFile(path);
        if (handle == 0) {
            throw new RuntimeException("Failed to load tokenizer from file: " + path);
        }
        return new Tokenizer(handle);
    }

    /**
     * Encode text into tokens
     * @param text The text to encode
     * @param addSpecialTokens Whether to add special tokens
     * @return An Encoding object
     */
    public Encoding encode(String text, boolean addSpecialTokens) {
        long encodingHandle = nativeEncode(handle, text, addSpecialTokens);
        if (encodingHandle == 0) {
            throw new RuntimeException("Failed to encode text");
        }
        return new Encoding(encodingHandle);
    }

    /**
     * Decode token IDs back to text
     * @param ids Array of token IDs
     * @param skipSpecialTokens Whether to skip special tokens
     * @return The decoded text
     */
    public String decode(int[] ids, boolean skipSpecialTokens) {
        return nativeDecode(handle, ids, skipSpecialTokens);
    }

    /**
     * Free the tokenizer resources
     */
    @Override
    public void close() {
        if (handle != 0) {
            nativeFree(handle);
            handle = 0;
        }
    }

    @Override
    protected void finalize() throws Throwable {
        close();
        super.finalize();
    }

    // Native method declarations
    private static native long nativeFromPretrained(String name, String revision, String token);
    private static native long nativeFromFile(String path);
    private static native long nativeEncode(long handle, String text, boolean addSpecialTokens);
    private static native String nativeDecode(long handle, int[] ids, boolean skipSpecialTokens);
    private static native void nativeFree(long handle);
}

File: examples/kotlin/src/main/kotlin/com/tokenizers/Tokenizer.kt
package com.tokenizers

class Tokenizer private constructor(private val handle: Long) : AutoCloseable {
    
    companion object {
        init {
            System.loadLibrary("tokenizers_jni")
        }
        
        @JvmStatic
        fun fromPretrained(name: String, revision: String? = null, token: String? = null): Tokenizer {
            val handle = nativeFromPretrained(name, revision, token)
            if (handle == 0L) {
                throw RuntimeException("Failed to create tokenizer from pretrained model: $name")
            }
            return Tokenizer(handle)
        }
        
        @JvmStatic
        fun fromFile(path: String): Tokenizer {
            val handle = nativeFromFile(path)
            if (handle == 0L) {
                throw RuntimeException("Failed to create tokenizer from file: $path")
            }
            return Tokenizer(handle)
        }
        
        @JvmStatic
        private external fun nativeFromPretrained(name: String, revision: String?, token: String?): Long
        
        @JvmStatic
        private external fun nativeFromFile(path: String): Long
        
        @JvmStatic
        private external fun nativeEncode(handle: Long, text: String, addSpecialTokens: Boolean): Long
        
        @JvmStatic
        private external fun nativeDecode(handle: Long, ids: IntArray, skipSpecialTokens: Boolean): String?
        
        @JvmStatic
        private external fun nativeFree(handle: Long)
    }
    
    fun encode(text: String, addSpecialTokens: Boolean = true): Encoding {
        val encodingHandle = nativeEncode(handle, text, addSpecialTokens)
        if (encodingHandle == 0L) {
            throw RuntimeException("Failed to encode text: $text")
        }
        return Encoding(encodingHandle)
    }
    
    fun decode(ids: IntArray, skipSpecialTokens: Boolean = true): String {
        return nativeDecode(handle, ids, skipSpecialTokens)
            ?: throw RuntimeException("Failed to decode token IDs")
    }
    
    override fun close() {
        if (handle != 0L) {
            nativeFree(handle)
        }
    }
}